validation: do not wipe utxo cache for stats/scans/snapshots

l0rinc opened the pull request:
Revival of https://github.com/bitcoin/bitcoin/pull/30610#issuecomment-3432564955 with the remaining comments applied on top

> Since #28280, the cost of a non-wiping sync of the UTXO cache is only proportional to the number of dirty entries, rather than proportional to the size of the entire cache. Because of that, there is no reason to perform a wiping flush in case the contents of the cache is still useful.
> 
> Split the `FlushStateMode::ALWAYS` mode into a FORCE_SYNC (non-wiping) and a FORCE_FLUSH (wiping), and then use the former in `scantxoutset`, `gettxoutsetinfo`, snapshot creation.

(slightly updated after #30214)

sipa committed:
validation: do not wipe utxo cache for stats/scans/snapshots

Since #28280, the cost of a non-wiping sync of the UTXO cache is only proportional to the number of dirty entries, rather than proportional to the size of the entire cache. Because of that, there is no reason to perform a wiping flush in case the contents of the cache is still useful.

Split the FlushStateMode::ALWAYS mode into a FORCE_SYNC (non-wiping) and a FORCE_FLUSH (wiping), and then use the former in scantxoutset, gettxoutsetinfo, snapshot creation.

Co-authored-by: l0rinc <pap.lorinc@gmail.com>
Co-authored-by: cedwies <141683552+cedwies@users.noreply.github.com>

cedwies commented on src/rpc/blockchain.cpp:
```diff
@@ -1044,7 +1044,7 @@ static RPCHelpMan gettxoutsetinfo()
     ChainstateManager& chainman = EnsureChainman(node);
     Chainstate& active_chainstate = chainman.ActiveChainstate();
-    active_chainstate.ForceFlushStateToDisk();
+    active_chainstate.ForceFlushStateToDisk(/*wipe_cache=*/false);
```
Did you consider using `active_chainstate.ForceFlushStateToDisk(/*wipe_cache=*/false);` conditionally depending on whether we use the index path or not? If we use the index path, isn't `ForceFlushStateToDisk()` unnecessary?

cedwies commented on src/test/fuzz/utxo_snapshot.cpp:
```diff
@@ -57,7 +57,7 @@ void sanity_check_snapshot()
     LOCK(cs_main);
     auto& cs{node.chainman->ActiveChainstate()};
-    cs.ForceFlushStateToDisk();
+    cs.ForceFlushStateToDisk(/*wipe_cache=*/true);
```
Why are we wiping the cache here? We only compute stats from `CoinsDB()`, not the in-memory cache, so a non-wiping flush (`wipe_cache=false`) should be enough for consistency (?)

l0rinc replied:
Good point, changed to `false`, added you as coauthor, thanks.

cedwies reviewed (commented):
Approach ACK

Do `ForceFlushToDisk(...)`/`FlushToDisk(...)` require `cs_main` to be held?
Most call sites hold it during shutdown flushes, but the function itself doesn’t assert or document that (no `AssertLockHeld(cs_main)`). If this is correct, I think a short comment can help that a Lock is not strictly needed.

l0rinc committed:
refactor: rename `FlushStateMode::ALWAYS` to `FORCE_FLUSH`

This prepares the addition of `FORCE_SYNC`.

`empty_cache` in `FlushStateToDisk` was moved up to be reusable and `FlushStateMode::FORCE_FLUSH` was used as a placeholder before we properly split the two new states.
`log_utxocache_flush.py` was regenerated and the alignment adjusted for the wider `FlushStateMode` values.

Co-authored-by: Pieter Wuille <pieter@wuille.net>
Co-authored-by: optout <13562139+optout21@users.noreply.github.com>

l0rinc commented on src/validation.cpp:
```diff
@@ -2909,10 +2909,10 @@ bool Chainstate::FlushStateToDisk(
     BlockValidationState state;
-    if (!this->FlushStateToDisk(state, FlushStateMode::ALWAYS)) {
+    if (!this->FlushStateToDisk(state, wipe_cache ? FlushStateMode::FORCE_FLUSH : FlushStateMode::FORCE_SYNC)) {
         LogPrintf("%s: failed to flush state (%s)\n", __func__, state.ToString());
```
We could, but I'm not touching that line, it's complicated enough as it is.

l0rinc commented:
Thanks @optout21 & @cedwies, [split](https://github.com/bitcoin/bitcoin/compare/b92ee755109d9eec70e25eb366f133d102687a6f..a59aacefd1a0cdb448c8f847665a5809f9897605) the change to 2 commits, one just renames the old value and adjusts the docs and scripts, the second introduces the new value.
I'm not sure about the locking, let's investigate that as a follow-up.

optout21 commented on contrib/tracing/README.md:
```diff
@@ -246,10 +246,10 @@ $ python3 contrib/tracing/log_utxocache_flush.py $(pidof bitcoind)
-637657          ALWAYS     122320          17124.80 kB     False
-81349           ALWAYS     0               1383.49 kB      False
+Duration (µs)   Mode         Coins Count     Memory Usage    Flush for Prune
+2556340         IF_NEEDED    2899141         394844.34 kB    False
```
I was wondering how tracing is affected by this change. The trace includes the mode, and it should display it fine, so the trace should be OK.

It would be nice to check how the trace looks for a FORCE_SYNC case as well; I guess that can be triggered by an RPC call to the running node.

optout21 replied:
> This is the actual output of running the script, looks like it wasn't updated in a while

I think that the scripts only enables the trace and listens to it, but it does not trigger a particular use case.
If trace outputs are displayed with these changes, that should be sufficient check (i.e., tracing a FORCE_SYNC is not crucial).

l0rinc commented:
rebased after #30595

cedwies commented:
utACK d442598

cedwies reviewed (commented):
reACK [14abac1](https://github.com/bitcoin/bitcoin/commit/14abac10f2ebd93367c91dde7586a35456ef9d45)

l0rinc commented:
Non-trivial rebase after https://github.com/bitcoin/bitcoin/pull/30214

Changes: 
* The new `chainstate_write_tests.cpp` also needed renaming in the first commit;
* `ForceFlushStateToDisk(bool wipe_cache = true)` is default-true now to minimize the diff.

cedwies commented:
checked the changes, ACK [11fe88a](https://github.com/bitcoin/bitcoin/commit/11fe88a327d8f8bd03b752686edb985ff2473ceb)

sedited reviewed (approved):
ACK c6ca2b85a3e6e73674e210aee4ed69c4af2848e4

Reproduced the diff from #30610.

achow101 commented:
ACK c6ca2b85a3e6e73674e210aee4ed69c4af2848e4
